### DISCUSSION TITLE
1 concurrent consumer vs N concurrent consumers

### DISCUSSION THREAD

**Discussion by pabermod** [rabbitmq/rabbitmq-dotnet-client#1306] [Q&A]
Hi, I have a doubt about which is the best approach for concurret consuming messages within the same client.

My scenario is: 
- Consumer receives the message
- Consumer process the message
- Consumer acks the message

I have tested that I could have for example a prefetch of 5 and I can process and ack each message inside a Task, so they will be processed concurrently. (Probably its not a good idea to create as many tasks as the preftech count in case is a big number, I could configure the grade of parallelism with a separate config key).

The other approach is to have another consumer. In this case, is it a good idea to share the same channel? (if I have a timeout while processing in one of the consumers it will close the channel affecting the other consumer....)

Which would be the best approach?

Thanks!

**Comment by michaelklishin** [comment]
Multiple consumers per channel rarely make things easier but if you think they do in your case, in terms of concurrent application code, you can use, say, 4-16 consumers per channel. It won't matter as far as RabbitMQ is concerned.

You can try N channels with 1 consumer each, that can yield a small throughput gain on nodes with a lot of cores available.

**Reply by pabermod** [reply]
Hi, in my case I open a channel for each queue I have. I need to be able to set a different concurrency degree for each queue. So I think the best approach might be having one consumer and one semaphore per queue and consume each message inside a Task. Also the prefetch for each channel must equals or greater to the concurrency degree

I also need to be able to change the concurrency degree on-the-fly... After doing some testings I see that if a change the preftech of the channel after creating the consumer, the consumer is still using the old prefetch (But the channel prefetch is indeed changed as a I cansee in the management interface). To apply the new prefetch to the consumer I have to manually cancel it and create the consumer again... I think this behaviour maybe is a bug?

**Reply by pabermod** [reply]
I also see that I cannot change ConsumerDispatchConcurrency on the fly because it can only be set in the connection factory

**Reply by lukebakken** [reply]
> After doing some testings I see that if a change the preftech of the channel after creating the consumer, the consumer is still using the old prefetch (But the channel prefetch is indeed changed as a I cansee in the management interface).

RabbitMQ is probably using the new prefetch, but the consumer object just may not be updated with the new value. If you could provide a small test case to demonstrate the issue that would be great.

**Reply by pabermod** [reply]
[ConsumerPrefetchUpdate.zip](https://github.com/rabbitmq/rabbitmq-dotnet-client/files/11029913/ConsumerPrefetchUpdate.zip)

Hi, I attach a basic example. Will update prefetech from 1 to 4 after 7 seconds. Each message processing is executed in a Task and takes 5 seconds.

The workaround where I manually cancel and create the consumer again is commented

**Reply by lukebakken** [reply]
Thanks, I added the code here: https://github.com/lukebakken/rabbitmq-dotnet-client-1306

**Reply by lukebakken** [reply]
Relevant discussion: https://groups.google.com/g/rabbitmq-users/c/uVb5jUX-KDw

**Reply by lukebakken** [reply]
@pabermod the docs for consumer prefetch state that the QoS value applies to _new_ consumers on a channel - https://www.rabbitmq.com/consumer-prefetch.html

So that means that you can't change it dynamically. You have to cancel the consumer, change the prefetch, and start a new consumer.

Your code is also sharing `IModel` instances among threads which will probably work for acks but isn't recommended - https://www.rabbitmq.com/dotnet-api-guide.html#concurrency

See this commit: https://github.com/lukebakken/rabbitmq-dotnet-client-1306/commit/a0b5942243cbddaeb43c75af12cb27097cd2e27a

**Reply by pabermod** [reply]
Hi, thanks for your response. So it's better to use separate `IModel` instances for each concurrent consumer? The problem I see is that I will always have N channels open per queue in the same connection. As the natural state of a queue is to be empty, how will affect performance to have N channels opened per queue instead of 1?

Edit: I think a misunderstood you, do you mean is better to create N consumers in the same channel instead of my example?

My problem is that I cannot use ConsumerDispatchConcurrency of connection factory, because I need to be able to update the concurrency level on the fly

**Reply by lukebakken** [reply]
> because I need to be able to update the concurrency level on the fly

You have to stop and start a consumer for that. It's how RabbitMQ works.

Just remember that the best practice is to _not_ share `IModel` instances among threads. I can't comment on the *best* way to not do that for your use-case - you'll have to benchmark that yourself.
